#Importaciones

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import datetime
from datetime import datetime
import re

In [2]:
from google.colab import drive
drive.mount('/content/drive') 

Mounted at /content/drive


In [3]:
db_train=pd.read_csv('/content/drive/MyDrive/DATASCIENCE/Practico/Practico 3/train (3).csv')

In [4]:
metadata=pd.read_csv('/content/drive/MyDrive/DATASCIENCE/Practico/Practico 3/metadata.csv',sep=';')

#Depuracion de datos

##metadata

In [5]:
metadata.isnull().sum()

asset_id                  0
content_id               21
title                     0
reduced_title             0
episode_title          4147
show_type                 4
released_year             0
country_of_origin         4
category                  0
keywords                  2
description               2
reduced_desc              0
cast_first_name        8732
credits_first_name    12554
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###content_id

In [6]:
metadata[metadata['content_id'].isnull()]

asset_id  content_id  ...          start_vod_date            end_vod_date
1540      15032         NaN  ...  2019-07-22T00:00:00.0Z  2020-12-14T23:59:59.0Z
1541      22963         NaN  ...  2019-07-22T00:00:00.0Z  2020-12-14T23:59:59.0Z
1542      32637         NaN  ...  2019-07-22T00:00:00.0Z  2020-12-14T23:59:59.0Z
1543      21822         NaN  ...  2019-07-22T00:00:00.0Z  2020-12-14T23:59:59.0Z
1544      17157         NaN  ...  2019-07-22T00:00:00.0Z  2020-12-14T23:59:59.0Z
1545      24583         NaN  ...  2019-07-22T00:00:00.0Z  2020-12-14T23:59:59.0Z
1546      27036         NaN  ...  2019-07-22T00:00:00.0Z  2020-12-14T23:59:59.0Z
18197      2046         NaN  ...  2019-07-22T00:00:00.0Z  2025-07-22T23:59:00.0Z
18198      1105         NaN  ...  2019-07-22T00:00:00.0Z  2025-07-22T23:59:00.0Z
18199     28435         NaN  ...  2019-07-22T00:00:00.0Z  2025-07-22T23:59:00.0Z
18200     17563         NaN  ...  2019-07-22T00:00:00.0Z  2025-07-22T23:59:00.0Z
18201     18599         NaN  ...  2019-07-22T00:00:00.0Z  2025-07-22T23:59:00.0Z
18202     27733         NaN  ...  2019-07-22T00:00:00.0Z  2025-07-22T23:59:00.0Z
18203     25592         NaN  ...  2019-07-22T00:00:00.0Z  2025-07-22T23:59:00.0Z
22895     15861         NaN  ...  2020-12-15T00:00:00.0Z  2025-07-22T23:59:00.0Z
22896      3938         NaN  ...  2020-12-15T00:00:00.0Z  2025-07-22T23:59:00.0Z
22897     12983         NaN  ...  2020-12-15T00:00:00.0Z  2025-07-22T23:59:00.0Z
22898     16881         NaN  ...  2020-12-15T00:00:00.0Z  2025-07-22T23:59:00.0Z
22899     19462         NaN  ...  2020-12-15T00:00:00.0Z  2025-07-22T23:59:00.0Z
22900     28382         NaN  ...  2020-12-15T00:00:00.0Z  2025-07-22T23:59:00.0Z
22901      6893         NaN  ...  2020-12-15T00:00:00.0Z  2025-07-22T23:59:00.0Z

[21 rows x 30 columns]

Se puede apreciar que estos casos son tutoriales/decodificador flow, por lo que no consideramos que sean contenidos recomendables y seran quitados.

In [7]:
metadata=metadata.drop(metadata[metadata['category']=='Tutoriales/Decodificador Flow'].index,axis=0)

In [8]:
metadata.shape

(33123, 30)

In [9]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title          4126
show_type                 4
released_year             0
country_of_origin         4
category                  0
keywords                  2
description               2
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###episode_title

In [10]:
metadata[metadata['episode_title'].isnull()]

asset_id  content_id  ...          start_vod_date            end_vod_date
1         24940         1.0  ...  2017-12-15T00:00:00.0Z  2022-12-14T23:59:59.0Z
2         21939         2.0  ...  2018-01-25T00:00:00.0Z  2020-12-01T23:59:59.0Z
3          9005         3.0  ...  2018-05-27T00:00:00.0Z  2021-04-30T23:59:59.0Z
4          7391         4.0  ...  2019-05-02T00:00:00.0Z  2020-12-31T23:59:59.0Z
10        21684         9.0  ...  2020-01-03T00:00:00.0Z  2021-12-31T23:59:59.0Z
...         ...         ...  ...                     ...                     ...
33118     11656      4367.0  ...  2021-03-29T00:00:00.0Z  2021-05-03T23:59:59.0Z
33123     11482      4368.0  ...  2021-03-29T00:00:00.0Z  2021-05-27T23:59:59.0Z
33124     11218      4369.0  ...  2021-03-31T00:00:00.0Z  2021-06-30T23:59:00.0Z
33132     27690      4370.0  ...  2021-05-10T00:00:00.0Z  2021-06-23T23:59:59.0Z
33137     19624      4371.0  ...  2021-05-06T00:00:00.0Z  2023-05-06T23:59:59.0Z

[4126 rows x 30 columns]

para este caso simplemente vamos a copiar el titulo reduced_title

In [11]:
metadata.loc[metadata['episode_title'].isnull(),'episode_title']=metadata[metadata['episode_title'].isnull()]['reduced_title']

In [12]:
metadata.head()

asset_id  content_id  ...          start_vod_date            end_vod_date
0     15188         0.0  ...  2017-12-01T00:00:00.0Z  2020-12-01T23:59:59.0Z
1     24940         1.0  ...  2017-12-15T00:00:00.0Z  2022-12-14T23:59:59.0Z
2     21939         2.0  ...  2018-01-25T00:00:00.0Z  2020-12-01T23:59:59.0Z
3      9005         3.0  ...  2018-05-27T00:00:00.0Z  2021-04-30T23:59:59.0Z
4      7391         4.0  ...  2019-05-02T00:00:00.0Z  2020-12-31T23:59:59.0Z

[5 rows x 30 columns]

In [13]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 4
released_year             0
country_of_origin         4
category                  0
keywords                  2
description               2
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###show_type

In [14]:
metadata[metadata['show_type'].isnull()]

asset_id  content_id  ...          start_vod_date            end_vod_date
3187      14004         9.0  ...  2020-11-20T00:00:00.0Z  2020-12-14T23:59:59.0Z
6421      11154       948.0  ...  2020-12-02T00:00:00.0Z  2021-01-01T23:59:00.0Z
7674      27728         9.0  ...  2020-12-15T00:00:00.0Z  2022-11-20T23:59:00.0Z
32927     24509      4315.0  ...  2021-03-24T00:00:00.0Z  2022-01-10T23:59:00.0Z

[4 rows x 30 columns]

In [15]:
metadata[metadata['content_id']==9.0]

asset_id  content_id  ...          start_vod_date            end_vod_date
10        21684         9.0  ...  2020-01-03T00:00:00.0Z  2021-12-31T23:59:59.0Z
24        28703         9.0  ...  2020-07-04T00:00:00.0Z  2021-02-28T23:59:00.0Z
3187      14004         9.0  ...  2020-11-20T00:00:00.0Z  2020-12-14T23:59:59.0Z
3331      14049         9.0  ...  2020-11-20T00:00:00.0Z  2020-12-14T23:59:59.0Z
7673      32237         9.0  ...  2020-12-15T00:00:00.0Z  2022-11-20T23:59:00.0Z
7674      27728         9.0  ...  2020-12-15T00:00:00.0Z  2022-11-20T23:59:00.0Z
29046     13209         9.0  ...  2020-11-20T00:00:00.0Z  2022-11-20T23:59:00.0Z
29047     15413         9.0  ...  2020-11-20T00:00:00.0Z  2022-11-20T23:59:00.0Z

[8 rows x 30 columns]

In [16]:
metadata[metadata['content_id']==948.0]

asset_id  content_id  ...          start_vod_date            end_vod_date
6421     11154       948.0  ...  2020-12-02T00:00:00.0Z  2021-01-01T23:59:00.0Z

[1 rows x 30 columns]

In [17]:
metadata[metadata['content_id']==4315.0	]

asset_id  content_id  ...          start_vod_date            end_vod_date
32927     24509      4315.0  ...  2021-03-24T00:00:00.0Z  2022-01-10T23:59:00.0Z

[1 rows x 30 columns]

Dado que el contenido id marca como 'Pelicula' al resto de los clasificados vamos a asignarle este valor.
y en el caso de l contenido 11154	y 24509 al ser unicos, vamos a tomar el mismo criterio.

In [18]:
metadata.loc[metadata['show_type'].isnull(),'show_type']='Película'

In [19]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         4
category                  0
keywords                  2
description               2
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###country_of_origin

In [20]:
metadata[metadata['country_of_origin'].isnull()]

asset_id  content_id  ...          start_vod_date            end_vod_date
7287      31314      1124.0  ...  2018-05-03T00:00:00.0Z  2021-01-16T23:59:59.0Z
7320      22004      1124.0  ...  2018-12-19T00:00:00.0Z  2021-01-16T23:59:00.0Z
7321       9452      1124.0  ...  2018-12-26T00:00:00.0Z  2021-01-16T23:59:00.0Z
17406      8946      2264.0  ...  2019-04-03T00:00:00.0Z  2022-03-31T23:59:00.0Z

[4 rows x 30 columns]

In [21]:
metadata[metadata['content_id']==1124.0]

asset_id  content_id  ...          start_vod_date            end_vod_date
7283      20070      1124.0  ...  2018-04-04T00:00:00.0Z  2021-01-16T23:59:59.0Z
7284       5900      1124.0  ...  2018-04-11T00:00:00.0Z  2021-01-16T23:59:59.0Z
7285      12651      1124.0  ...  2018-04-18T00:00:00.0Z  2021-01-16T23:59:59.0Z
7286       3304      1124.0  ...  2018-04-26T00:00:00.0Z  2021-01-16T23:59:59.0Z
7287      31314      1124.0  ...  2018-05-03T00:00:00.0Z  2021-01-16T23:59:59.0Z
7288      24603      1124.0  ...  2018-05-09T00:00:00.0Z  2021-01-16T23:59:59.0Z
7289       5512      1124.0  ...  2018-05-17T00:00:00.0Z  2021-01-16T23:59:59.0Z
7290      11866      1124.0  ...  2018-05-23T00:00:00.0Z  2021-01-16T23:59:59.0Z
7291      31286      1124.0  ...  2018-05-30T00:00:00.0Z  2021-01-16T23:59:59.0Z
7292      22327      1124.0  ...  2018-06-13T00:00:00.0Z  2021-01-16T23:59:59.0Z
7293       5307      1124.0  ...  2018-06-20T00:00:00.0Z  2021-01-16T23:59:59.0Z
7294       3294      1124.0  ...  2018-06-06T00:00:00.0Z  2021-01-16T23:59:59.0Z
7295      28693      1124.0  ...  2018-06-27T00:00:00.0Z  2021-01-16T23:59:59.0Z
7296      27618      1124.0  ...  2018-07-04T00:00:00.0Z  2021-01-16T23:59:59.0Z
7297      25314      1124.0  ...  2018-07-11T00:00:00.0Z  2021-01-16T23:59:59.0Z
7298      10338      1124.0  ...  2018-07-18T00:00:00.0Z  2021-01-16T23:59:59.0Z
7299      25084      1124.0  ...  2018-07-25T00:00:00.0Z  2021-01-16T23:59:59.0Z
7300      14848      1124.0  ...  2018-09-06T00:00:00.0Z  2021-01-16T23:59:00.0Z
7301       4152      1124.0  ...  2018-09-12T00:00:00.0Z  2021-01-16T23:59:59.0Z
7302       6911      1124.0  ...  2018-09-19T00:00:00.0Z  2021-01-16T23:59:59.0Z
7303       2778      1124.0  ...  2018-03-28T00:00:00.0Z  2021-01-16T23:59:59.0Z
7304      10534      1124.0  ...  2018-08-01T00:00:00.0Z  2021-01-16T23:59:59.0Z
7305       4022      1124.0  ...  2018-09-26T00:00:00.0Z  2021-01-16T23:59:59.0Z
7306       1560      1124.0  ...  2018-10-03T00:00:00.0Z  2021-01-16T23:59:59.0Z
7307      13340      1124.0  ...  2018-10-10T00:00:00.0Z  2021-01-16T23:59:59.0Z
7308      19377      1124.0  ...  2018-10-17T00:00:00.0Z  2021-01-16T23:59:59.0Z
7309      21177      1124.0  ...  2018-10-24T00:00:00.0Z  2021-01-16T23:59:59.0Z
7310      21884      1124.0  ...  2018-10-31T00:00:00.0Z  2021-01-16T23:59:59.0Z
7311      31805      1124.0  ...  2018-11-07T00:00:00.0Z  2021-01-16T23:59:59.0Z
7312      32503      1124.0  ...  2018-11-14T00:00:00.0Z  2021-01-16T23:59:59.0Z
7313      21545      1124.0  ...  2018-11-21T00:00:00.0Z  2021-01-16T23:59:59.0Z
7314      31249      1124.0  ...  2018-11-28T00:00:00.0Z  2021-01-16T23:59:59.0Z
7315       9999      1124.0  ...  2018-12-05T00:00:00.0Z  2021-01-16T23:59:59.0Z
7316      16781      1124.0  ...  2018-12-12T00:00:00.0Z  2021-01-16T23:59:59.0Z
7317       1514      1124.0  ...  2019-01-02T00:00:00.0Z  2021-01-16T23:59:59.0Z
7318       7356      1124.0  ...  2019-01-09T00:00:00.0Z  2021-01-16T23:59:59.0Z
7319      18756      1124.0  ...  2019-01-16T00:00:00.0Z  2021-01-16T23:59:59.0Z
7320      22004      1124.0  ...  2018-12-19T00:00:00.0Z  2021-01-16T23:59:00.0Z
7321       9452      1124.0  ...  2018-12-26T00:00:00.0Z  2021-01-16T23:59:00.0Z
17697     16885      1124.0  ...  2019-05-08T00:00:00.0Z  2021-05-08T23:59:00.0Z
17794     24165      1124.0  ...  2019-05-15T00:00:00.0Z  2021-05-15T23:59:00.0Z
17877     16826      1124.0  ...  2019-05-22T00:00:00.0Z  2021-05-22T23:59:00.0Z
17959      9691      1124.0  ...  2019-05-29T00:00:00.0Z  2021-05-29T23:59:00.0Z
17999      5363      1124.0  ...  2019-06-05T00:00:00.0Z  2021-06-05T23:59:00.0Z
18035      5838      1124.0  ...  2019-06-12T00:00:00.0Z  2021-06-12T23:59:00.0Z
18112      8993      1124.0  ...  2019-06-26T00:00:00.0Z  2021-06-26T23:59:00.0Z

[46 rows x 30 columns]

In [22]:
metadata[metadata['content_id']==2264.0]

asset_id  content_id  ...          start_vod_date            end_vod_date
17406      8946      2264.0  ...  2019-04-03T00:00:00.0Z  2022-03-31T23:59:00.0Z

[1 rows x 30 columns]

Al ver los content_id, presumimos que los contenidos son de US, menos el 8946 que investigando vemos que se trata de origen Canadiense.

In [23]:
metadata.loc[metadata['asset_id']==8946,'country_of_origin']='CA'

In [24]:
metadata.loc[metadata['country_of_origin'].isnull(),'country_of_origin']='US'

In [25]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         0
category                  0
keywords                  2
description               2
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

In [26]:
metadata[metadata['keywords'].isnull()]

asset_id  content_id  ...          start_vod_date            end_vod_date
29761     25451      3914.0  ...  2021-02-07T00:00:00.0Z  2021-06-30T23:59:00.0Z
30564     27065      1175.0  ...  2021-02-20T00:00:00.0Z  2021-04-06T23:59:00.0Z

[2 rows x 30 columns]

In [27]:
metadata[metadata['content_id']==1175.0]

asset_id  content_id  ...          start_vod_date            end_vod_date
7635      18103      1175.0  ...  2021-01-07T00:00:00.0Z  2021-01-21T23:59:00.0Z
7971      19846      1175.0  ...  2021-01-14T00:00:00.0Z  2021-01-28T23:59:00.0Z
8763        569      1175.0  ...  2021-01-22T00:00:00.0Z  2021-02-04T23:59:00.0Z
8781      10675      1175.0  ...  2020-09-05T00:00:00.0Z  2021-02-05T23:59:59.0Z
9152       5142      1175.0  ...  2021-01-29T00:00:00.0Z  2021-02-10T23:59:00.0Z
9205      24351      1175.0  ...  2020-09-12T00:00:00.0Z  2021-02-12T23:59:59.0Z
10077     29720      1175.0  ...  2020-09-26T00:00:00.0Z  2021-02-26T23:59:59.0Z
11403      3352      1175.0  ...  2020-09-03T00:00:00.0Z  2021-03-03T23:59:59.0Z
11709     18768      1175.0  ...  2020-09-10T00:00:00.0Z  2021-03-10T23:59:59.0Z
12381      8877      1175.0  ...  2020-09-17T00:00:00.0Z  2021-03-17T23:59:59.0Z
12382     14754      1175.0  ...  2020-09-24T00:00:00.0Z  2021-03-24T23:59:59.0Z
12697     24564      1175.0  ...  2021-02-06T00:00:00.0Z  2021-03-24T23:59:00.0Z
30148     15356      1175.0  ...  2021-02-13T00:00:00.0Z  2021-03-31T23:59:59.0Z
30564     27065      1175.0  ...  2021-02-20T00:00:00.0Z  2021-04-06T23:59:00.0Z
30826     12080      1175.0  ...  2021-02-27T00:00:00.0Z  2021-04-13T23:59:00.0Z
31765     14030      1175.0  ...  2021-03-04T00:00:00.0Z  2021-04-21T23:59:00.0Z

[16 rows x 30 columns]

Vamos a usar como palabras clave 'Niños,Ingenioso' para el contenido 27065



In [28]:
metadata.loc[metadata['asset_id']==27065,'keywords']='Niños,Ingenioso'

In [29]:
metadata.loc[metadata['title']=='Bourne: El Ultimátum']['keywords']

29758    Agentes,Nueva York,Taquilleras,Espionaje
Name: keywords, dtype: object

Para el caso de 25451 vamos a copiar keywords de una pelicula de la misma saga, la 21097

In [30]:
metadata.loc[metadata['asset_id']==25451,'keywords']='Agentes,Nueva York,Taquilleras,Espionaje'

In [31]:
metadata.loc[metadata['asset_id']==25451,'keywords']

29761    Agentes,Nueva York,Taquilleras,Espionaje
Name: keywords, dtype: object

In [32]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         0
category                  0
keywords                  0
description               2
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###description

In [33]:
metadata[metadata['description'].isnull()]

asset_id  content_id  ...          start_vod_date            end_vod_date
16238      5514      2168.0  ...  2018-11-19T00:00:00.0Z  2021-03-31T23:59:59.0Z
16330     21755      2168.0  ...  2019-11-26T00:00:00.0Z  2021-03-31T23:59:59.0Z

[2 rows x 30 columns]

Para estos casos vamos a duplicar la descripcion reducida

In [34]:
metadata.loc[metadata['description'].isnull(),'description']=metadata[metadata['description'].isnull()]['reduced_desc']

In [35]:
metadata.loc[metadata['asset_id']==21755]

asset_id  content_id  ...          start_vod_date            end_vod_date
16330     21755      2168.0  ...  2019-11-26T00:00:00.0Z  2021-03-31T23:59:59.0Z

[1 rows x 30 columns]

In [36]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         0
category                  0
keywords                  0
description               0
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###cast_first_name

In [37]:
metadata[metadata['cast_first_name'].isnull()]

asset_id  content_id  ...          start_vod_date            end_vod_date
9         32482         8.0  ...  2019-10-09T00:00:00.0Z  2020-12-01T23:59:59.0Z
11        30574        10.0  ...  2020-03-02T00:00:00.0Z  2020-12-01T23:59:00.0Z
18        24600        17.0  ...  2020-03-02T00:00:00.0Z  2020-12-01T23:59:00.0Z
19        14972        18.0  ...  2020-03-29T00:00:00.0Z  2020-12-01T23:59:59.0Z
20        18155        19.0  ...  2020-05-06T00:00:00.0Z  2020-12-01T23:59:59.0Z
...         ...         ...  ...                     ...                     ...
33106     31039      1290.0  ...  2021-03-05T00:00:00.0Z  2021-05-05T23:59:00.0Z
33107     16797      1290.0  ...  2021-03-05T00:00:00.0Z  2021-05-05T23:59:00.0Z
33113     21955      1039.0  ...  2021-03-15T00:00:00.0Z  2021-05-15T23:59:00.0Z
33114      4823       154.0  ...  2021-03-29T00:00:00.0Z  2021-05-29T23:59:59.0Z
33137     19624      4371.0  ...  2021-05-06T00:00:00.0Z  2023-05-06T23:59:59.0Z

[8711 rows x 30 columns]

Lamentablemente esta columna la vamos a dejar sin tratar

In [38]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         0
category                  0
keywords                  0
description               0
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###credits_first_name

In [39]:
metadata[metadata['credits_first_name'].isnull()]

asset_id  content_id  ...          start_vod_date            end_vod_date
7          2472         7.0  ...  2019-12-02T00:00:00.0Z  2020-12-01T23:59:59.0Z
9         32482         8.0  ...  2019-10-09T00:00:00.0Z  2020-12-01T23:59:59.0Z
19        14972        18.0  ...  2020-03-29T00:00:00.0Z  2020-12-01T23:59:59.0Z
20        18155        19.0  ...  2020-05-06T00:00:00.0Z  2020-12-01T23:59:59.0Z
21        22721        20.0  ...  2020-05-19T00:00:00.0Z  2020-12-01T23:59:59.0Z
...         ...         ...  ...                     ...                     ...
33110     24551      3891.0  ...  2021-03-30T00:00:00.0Z  2021-06-01T23:59:00.0Z
33111     28170      1362.0  ...  2021-03-30T00:00:00.0Z  2021-06-30T23:59:00.0Z
33112     27267      1409.0  ...  2021-03-30T00:00:00.0Z  2022-01-15T23:59:00.0Z
33116     17733       933.0  ...  2021-03-11T00:00:00.0Z  2021-06-09T23:59:00.0Z
33121     10134      4046.0  ...  2021-03-30T00:00:00.0Z  2021-06-30T23:59:00.0Z

[12533 rows x 30 columns]

Esta columna tambien queda sin tratar

In [40]:
metadata.isnull().sum()

asset_id                  0
content_id                0
title                     0
reduced_title             0
episode_title             0
show_type                 0
released_year             0
country_of_origin         0
category                  0
keywords                  0
description               0
reduced_desc              0
cast_first_name        8711
credits_first_name    12533
run_time_min              0
audience                  1
made_for_tv               0
close_caption             0
sex_rating                0
violence_rating           0
language_rating           0
dialog_rating             0
fv_rating                 0
pay_per_view              0
pack_premium_1            0
pack_premium_2            0
create_date               0
modify_date               0
start_vod_date            0
end_vod_date              0
dtype: int64

###audience

In [41]:
metadata[metadata['audience'].isnull()]

asset_id  content_id  ...          start_vod_date            end_vod_date
14523     26900      2064.0  ...  2018-01-05T00:00:00.0Z  2029-12-31T23:59:00.0Z

[1 rows x 30 columns]

In [42]:
metadata[metadata['content_id']==2064.0]

asset_id  content_id  ...          start_vod_date            end_vod_date
14523     26900      2064.0  ...  2018-01-05T00:00:00.0Z  2029-12-31T23:59:00.0Z

[1 rows x 30 columns]

In [43]:
metadata['audience'].unique()

array(['General', 'Juvenil', 'Preescolar', 'Teens', 'Hombres', 'Familiar',
       'Niños', 'Mujeres', 'Gaming', 'NIños', nan], dtype=object)

Este caso es muy particular, se trata del capitulo 8 de la temporada 1 de una serie que no tiene el resto de capitulos. Vamos a clasificarla como de audiencia general.

In [44]:
metadata.loc[metadata['audience'].isnull(),'audience']='General'

In [243]:
metadata.to_csv('/content/drive/MyDrive/DATASCIENCE/Practico/Practico 3/metadatalimpia.csv', index=False)

##db_train

In [45]:
db_train.shape

(3657801, 7)

In [46]:
db_train.head()

customer_id  account_id  ...                tuneout  resume
0            0       90627  ...  2021-02-18 23:35:00.0       0
1            0       90627  ...  2021-03-25 00:01:00.0       0
2            1        3387  ...  2021-03-15 10:23:00.0       0
3            1        3387  ...  2021-03-15 11:18:00.0       1
4            1        3387  ...  2021-03-16 09:44:00.0       0

[5 rows x 7 columns]

In [47]:
db_train.isnull().sum()

customer_id     0
account_id      0
device_type    29
asset_id       22
tunein          0
tuneout         0
resume          0
dtype: int64

###device_type

In [48]:
db_train[db_train['device_type'].isnull()]

customer_id  account_id  ...                tuneout  resume
528885         18693      105629  ...  2021-01-17 17:07:00.0       0
666823         23439        6012  ...  2021-01-17 21:58:00.0       0
737730         25729       46974  ...  2021-01-13 11:33:00.0       0
737731         25729       46974  ...  2021-01-13 11:36:00.0       0
737732         25729       46974  ...  2021-01-13 12:36:00.0       0
931142         32196       52544  ...  2021-02-10 23:43:00.0       0
1020856        34976        6983  ...  2021-01-20 15:13:00.0       0
1668627        53378       71627  ...  2021-01-13 06:38:00.0       0
1785644        56361       74384  ...  2021-01-13 20:47:00.0       0
2038544        64195       82302  ...  2021-01-16 17:56:00.0       0
2038545        64195       82302  ...  2021-01-16 18:33:00.0       0
2038546        64195       82302  ...  2021-01-16 19:55:00.0       0
2038547        64195       82302  ...  2021-01-16 20:05:00.0       0
2038548        64195       82302  ...  2021-01-16 20:59:00.0       0
2038549        64195       82302  ...  2021-01-16 23:42:00.0       0
2038550        64195       82302  ...  2021-01-17 09:58:00.0       0
2038551        64195       82302  ...  2021-01-17 10:45:00.0       0
2731969        82236      101874  ...  2021-01-15 17:40:00.0       0
2740223        82422      102111  ...  2021-01-15 02:46:00.0       0
2740225        82422      102111  ...  2021-01-15 03:50:00.0       0
2993856        87982      108985  ...  2021-02-11 20:37:00.0       0
3025633        88889      110083  ...  2021-02-10 22:14:00.0       0
3025634        88889      110083  ...  2021-02-10 22:22:00.0       0
3025635        88889      110083  ...  2021-02-10 23:04:00.0       0
3144002        93905      100979  ...  2021-01-14 22:49:00.0       0
3254238        97749       14892  ...  2021-01-16 01:29:00.0       0
3254239        97749       14892  ...  2021-01-16 15:03:00.0       0
3352578       101276       17621  ...  2021-01-16 22:24:00.0       0
3635973       111507       25402  ...  2021-03-07 22:52:00.0       0

[29 rows x 7 columns]

Como no va a ser relevante el dispositivo desde el cual se conecta, vamos a ignorar estos nulos

###asset_id

In [49]:
db_train[db_train['asset_id'].isnull()]

customer_id  account_id  ...                tuneout  resume
106890          3746       28992  ...  2021-03-20 12:52:00.0       1
120175          4277       29416  ...  2021-03-16 00:36:00.0       0
146642          5174       30114  ...  2021-03-28 22:39:00.0       0
219038          7685       32112  ...  2021-03-20 21:55:00.0       0
249196          8809       33030  ...  2021-03-20 15:18:00.0       0
484894         17133       39757  ...  2021-03-22 16:51:00.0       0
726181         25327       46629  ...  2021-03-26 22:09:00.0       0
863794         29946       50616  ...  2021-03-24 22:09:00.0       0
1062022        36186       56086  ...  2021-03-17 18:58:00.0       0
1238162        41180       60401  ...  2021-03-12 17:52:00.0       0
1321333        43598       62625  ...  2021-03-23 12:45:00.0       0
1652989        52885       71182  ...  2021-03-21 01:03:00.0       0
1854385        58326       76218  ...  2021-03-13 13:21:00.0       0
1860974        58550       76429  ...  2021-03-17 19:30:00.0       0
2200219        68671       86969  ...  2021-03-25 18:36:00.0       0
2348268        72746       91253  ...  2021-03-27 03:38:00.0       0
2982344        87692      108646  ...  2021-03-18 00:07:00.0       0
3105229        92460       10784  ...  2021-03-12 13:36:00.0       0
3186447        95433       13058  ...  2021-03-15 20:26:00.0       0
3490722       106241       21345  ...  2021-03-23 15:58:00.0       0
3572923       109282       23657  ...  2021-03-20 01:46:00.0       0
3611436       110642       24742  ...  2021-03-18 21:14:00.0       1

[22 rows x 7 columns]

En este caso, dado que no tenemos el contenido que se vió, vamos a proceder a eliminar estas filas.

In [50]:
db_train=db_train.drop(db_train[db_train['asset_id'].isnull()].index,axis=0)

In [51]:
db_train.isnull().sum()

customer_id     0
account_id      0
device_type    29
asset_id        0
tunein          0
tuneout         0
resume          0
dtype: int64

#TRAINT-TEST

Para este ejercicio solo voy a tomar contenido visto como aquel en el que quedó registrado en la columna asset_id, omitiendo si la visualizacion del contenido fue casi nula, parcial o total.
 Es por esto que solo me voy a quedar con las columnas account_id, asset_id.
 Tambien para su posterior separacion en train y test vamos a conservar la columna tuneout.

In [52]:
db=db_train[['account_id','asset_id','tuneout']]

In [53]:
db.shape

(3657779, 3)

In [54]:
db.head()

account_id  asset_id                tuneout
0       90627   18332.0  2021-02-18 23:35:00.0
1       90627   24727.0  2021-03-25 00:01:00.0
2        3387     895.0  2021-03-15 10:23:00.0
3        3387     895.0  2021-03-15 11:18:00.0
4        3387   26062.0  2021-03-16 09:44:00.0

Lo siguiente es transformar asset_id en content_id

In [55]:
db_mod=pd.merge(db,metadata[['asset_id','content_id','end_vod_date','category','keywords']],how='left')

In [56]:
db_mod.head()

account_id  ...                                           keywords
0       90627  ...  Sociedad,Tragicómica,Adicciones,Familia,Amigos...
1       90627  ...  Sociedad,Tragicómica,Adicciones,Familia,Amigos...
2        3387  ...  Crimen,Abusos,Mujeres,De Libros,Feminismo,Gold...
3        3387  ...  Crimen,Abusos,Mujeres,De Libros,Feminismo,Gold...
4        3387  ...                      30s,Espionaje,Historia,Europa

[5 rows x 7 columns]

In [57]:
db_mod=db_mod.drop(['asset_id'],axis=1)

In [58]:
db_mod.head()

account_id  ...                                           keywords
0       90627  ...  Sociedad,Tragicómica,Adicciones,Familia,Amigos...
1       90627  ...  Sociedad,Tragicómica,Adicciones,Familia,Amigos...
2        3387  ...  Crimen,Abusos,Mujeres,De Libros,Feminismo,Gold...
3        3387  ...  Crimen,Abusos,Mujeres,De Libros,Feminismo,Gold...
4        3387  ...                      30s,Espionaje,Historia,Europa

[5 rows x 6 columns]

##Cambio de columna Time


In [59]:
db_mod['end_vod_date']=pd.to_datetime(db_mod.end_vod_date)

In [60]:
db_mod['tuneout']=pd.to_datetime(db_mod.tuneout)

In [61]:
db_mod['end_vod_date'] = db_mod['end_vod_date'].apply(lambda x: x.replace(tzinfo=None).date())

In [62]:
db_mod['tuneout'] = db_mod['tuneout'].apply(lambda x: x.replace(tzinfo=None).date())

##Guardamos el archivo

In [63]:
db_mod.to_csv('/content/drive/MyDrive/DATASCIENCE/Practico/Practico 3/db_mod.csv', index=False)

##Quitamos contenido.
Vamos a eliminar todo el contenido que ya no a va a estar disponible.

In [64]:
db_mod.shape

(3657779, 6)

In [65]:
type(db_mod['end_vod_date'][0])

datetime.date

In [66]:
db_mod=db_mod[db_mod['end_vod_date']>datetime(2021, 3, 1).date()]

In [67]:
db_mod.shape

(3354829, 6)

In [68]:
min(db_mod['end_vod_date'])

datetime.date(2021, 3, 2)

In [69]:
db_mod=db_mod.drop('end_vod_date', axis=1)

In [70]:
db_mod

account_id  ...                                           keywords
0             90627  ...  Sociedad,Tragicómica,Adicciones,Familia,Amigos...
1             90627  ...  Sociedad,Tragicómica,Adicciones,Familia,Amigos...
2              3387  ...  Crimen,Abusos,Mujeres,De Libros,Feminismo,Gold...
3              3387  ...  Crimen,Abusos,Mujeres,De Libros,Feminismo,Gold...
4              3387  ...                      30s,Espionaje,Historia,Europa
...             ...  ...                                                ...
3657774        3386  ...                        Aventuras,Autos,Deportistas
3657775        3386  ...                        Aventuras,Autos,Deportistas
3657776        3386  ...                        Aventuras,Autos,Deportistas
3657777        3386  ...                        Aventuras,Autos,Deportistas
3657778        3386  ...                        Aventuras,Autos,Deportistas

[3354829 rows x 5 columns]

##Separacion entre train y test
Para esta separacion el criterio es por fecha, todo lo anterior a 1 de marzo de 2021 será train y lo posterior test.

In [71]:
db_train=db_mod[db_mod['tuneout']<=datetime(2021, 3, 1).date()]

In [72]:
db_test=db_mod[db_mod['tuneout']>datetime(2021, 3, 1).date()]

In [73]:
db_train.shape

(2072477, 5)

In [74]:
db_test.shape

(1282352, 5)

##Eliminamos tuneout

In [75]:
db_train=db_train.drop('tuneout',axis=1)

In [76]:
db_test=db_test.drop('tuneout',axis=1)

In [77]:
db_train.head()

account_id  ...                                           keywords
0       90627  ...  Sociedad,Tragicómica,Adicciones,Familia,Amigos...
6        3388  ...                                            Médicos
7        3388  ...                                            Médicos
8        3388  ...                                            Médicos
9        3388  ...                                            Médicos

[5 rows x 4 columns]

In [78]:
db_test.head()

account_id  ...                                           keywords
1       90627  ...  Sociedad,Tragicómica,Adicciones,Familia,Amigos...
2        3387  ...  Crimen,Abusos,Mujeres,De Libros,Feminismo,Gold...
3        3387  ...  Crimen,Abusos,Mujeres,De Libros,Feminismo,Gold...
4        3387  ...                      30s,Espionaje,Historia,Europa
5        3387  ...                      30s,Espionaje,Historia,Europa

[5 rows x 4 columns]

##Eliminamos duplicados
Vamos a eliminar los registros duplicados, estos registros se generaron dado que al unificar los capitulos por serie, un id que vio varios capitulos de una serie, va a figurar varias veces.

In [79]:
db_train.shape

(2072477, 4)

In [80]:
db_train=db_train.drop_duplicates(subset=['account_id', 'content_id'], keep='last')

In [81]:
db_train.shape

(598181, 4)

In [82]:
db_train['account_id'].nunique()

97984

In [83]:
db_test=db_test.drop_duplicates()

#Contenido mas popular

Aca vamos a colocar en la variable 'popular' el contenido mas visto. Luego recomendaremos estos contenidos a los usuarios nuevos en la funcion Recomendar.

In [119]:
populares=db_train['content_id'].value_counts().head(20).index.to_list()

In [120]:
populares

[2040.0,
 3806.0,
 3900.0,
 3598.0,
 2160.0,
 3381.0,
 3690.0,
 3210.0,
 3384.0,
 3550.0,
 2627.0,
 3711.0,
 3775.0,
 3382.0,
 2942.0,
 116.0,
 3716.0,
 2827.0,
 3712.0,
 724.0]

#Matriz Interaction 


In [86]:
db_train.head()

account_id  ...                                           keywords
0        90627  ...  Sociedad,Tragicómica,Adicciones,Familia,Amigos...
12        3388  ...                           Narcos,Nueva York,Crimen
30        3388  ...  Mujeres,Superación,Amigos,Tragicómica,Corporación
38        3388  ...  Mujeres,60s,Aviones,Viajes,Época,Melodrama,His...
48        3388  ...                                            Médicos

[5 rows x 4 columns]

In [87]:
ddbb=db_train.assign(miro=1)

In [88]:
interaction=ddbb.pivot(index='account_id',columns='content_id', values='miro')

In [89]:
interaction.shape

(97984, 3026)

In [90]:
interaction.fillna(value=0,inplace=True)

In [91]:
interaction

content_id  0.0     1.0     3.0     6.0     ...  4363.0  4364.0  4365.0  4366.0
account_id                                  ...                                
0              0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
2              0.0     0.0     0.0     1.0  ...     0.0     0.0     0.0     0.0
3              0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
4              0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
5              0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
...            ...     ...     ...     ...  ...     ...     ...     ...     ...
111503         0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
111507         0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
111543         0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
111572         0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0
111605         0.0     0.0     0.0     0.0  ...     0.0     0.0     0.0     0.0

[97984 rows x 3026 columns]

##indices de referencia

In [92]:
account_id = list(interaction.index)
user_dict = {}
counter = 0 
for i in account_id:
    user_dict[i] = counter
    counter += 1

In [93]:
content_id = list(interaction.columns)
item_dict = {}
counter = 0 
for i in content_id:
    item_dict[i] = counter
    counter += 1

##Converimos la matriz interaction

In [94]:
from scipy.sparse import csr_matrix

In [95]:
interaction_csr = csr_matrix(interaction.values)
interaction_csr

<97984x3026 sparse matrix of type '<class 'numpy.float64'>'
	with 598181 stored elements in Compressed Sparse Row format>

In [96]:
interaction_csr.shape

(97984, 3026)

#ENTRENAMOS EL MODELO

In [97]:
pip install lightfm

     |████████████████████████████████| 310 kB 5.3 MB/s 
  Created wheel for lightfm: filename=lightfm-1.16-cp37-cp37m-linux_x86_64.whl size=705379 sha256=5a9e9b4acd2245a601f92b3d1dc249f096ff9ef24570b903426b3d65f4ea77dd
  Stored in directory: /root/.cache/pip/wheels/f8/56/28/5772a3bd3413d65f03aa452190b00898b680b10028a1021914
Successfully built lightfm


In [98]:
from lightfm import LightFM

In [99]:
model = LightFM(loss='warp',no_components=10)


In [100]:
model=model.fit(interaction_csr,epochs=20)

In [101]:
user_x = user_dict[18]
n_users, n_items = interaction.shape
item_ids = np.arange(n_items)
#ACA GENERAMOS PREDICCIONES
preds = model.predict(user_ids=user_x, item_ids = item_ids)

In [102]:
preds

array([-2.6118438 , -1.7830757 , -2.478799  , ...,  0.03321559,
       -0.40076742,  0.03327411], dtype=float32)

In [103]:
scores = pd.Series(preds)
scores.index = interaction.columns
recomms_18 = list(pd.Series(scores.sort_values(ascending=False).index))[:10] #obtenemos las primeras 10

In [104]:
recomms_18

[3711.0,
 3210.0,
 3900.0,
 3206.0,
 1008.0,
 3381.0,
 3384.0,
 1070.0,
 3806.0,
 1495.0]

#Funcion Recomendar

In [118]:
def Recomendador(Unser_ID):
  #Esta funcion recibe un usuario id, lo busca en la base de datos y si lo encuentra le da una recomendacion de 40 contenidos recomendados,
  #caso contrario recomienda los 20 contenidos mas populares.
  
  if Unser_ID in list(interaction.index):
    
    user_x = user_dict[Unser_ID]
    n_users, n_items = interaction.shape
    item_ids = np.arange(n_items)
    preds = model.predict(user_ids=user_x, item_ids = item_ids)

    scores = pd.Series(preds)
    scores.index = interaction.columns
    Recomendadas = list(pd.Series(scores.sort_values(ascending=False).index))[:20] #obtenemos las primeras 10

    #chequeamos que el contenido no fue previamente visto
    vio=db_train[db_train['account_id']==2].content_id.unique()
    for i in vio:
      for j in Recomendadas:
        if i==j:
          Recomendadas.remove(i)


  else:
    Recomendadas=populares

  return Recomendadas

In [106]:
Recomendador(2)

[3900.0,
 3806.0,
 2040.0,
 1531.0,
 1971.0,
 2942.0,
 3863.0,
 3716.0,
 598.0,
 3382.0,
 718.0,
 2180.0,
 3726.0,
 1908.0,
 3690.0,
 2815.0,
 3210.0,
 3775.0,
 2901.0,
 592.0,
 1877.0,
 3712.0,
 1800.0,
 3711.0,
 3689.0,
 623.0,
 3364.0,
 2178.0,
 558.0,
 1176.0,
 604.0,
 3091.0,
 576.0,
 1495.0,
 1769.0,
 3048.0,
 3043.0,
 3402.0,
 3519.0]

#Evaluacion

In [215]:
tabla_recomendada=db_test[['account_id','content_id']]

In [216]:
tabla_recomendada.head()

account_id  content_id
1        90627      2040.0
2         3387      1983.0
4         3387       729.0
50        3388      2040.0
60        3388      3386.0

In [217]:
tabla_recomendada.shape

(362410, 2)

In [218]:
tabla_recomendada['account_id'].nunique()

86878

In [219]:
tabla_recomendada=tabla_recomendada.groupby('account_id', as_index=False).agg({"content_id": "unique"})

In [220]:
tabla_recomendada = pd.DataFrame(tabla_recomendada)

In [221]:
tabla_recomendada.head()

account_id                                         content_id
0           2                                            [433.0]
1           3  [1949.0, 2409.0, 4010.0, 3169.0, 3487.0, 3980....
2           4  [2314.0, 728.0, 4129.0, 2344.0, 2341.0, 513.0,...
3           5                                           [2259.0]
4           6           [3211.0, 2900.0, 3386.0, 4065.0, 3388.0]

In [210]:
tabla_recomendada.shape

(86878, 2)

In [222]:
Rec=[] #Genero una columna con todas recomendaciones populares.
for i in tabla_recomendada['account_id']:
  Rec.append(populares)

In [224]:
tabla_recomendada.shape

(86878, 2)

In [226]:
tabla_recomendada['Popular']=Rec

In [227]:
tabla_recomendada.head()

account_id  ...                                            Popular
0           2  ...  [2040.0, 3806.0, 3900.0, 3598.0, 2160.0, 3381....
1           3  ...  [2040.0, 3806.0, 3900.0, 3598.0, 2160.0, 3381....
2           4  ...  [2040.0, 3806.0, 3900.0, 3598.0, 2160.0, 3381....
3           5  ...  [2040.0, 3806.0, 3900.0, 3598.0, 2160.0, 3381....
4           6  ...  [2040.0, 3806.0, 3900.0, 3598.0, 2160.0, 3381....

[5 rows x 3 columns]

In [131]:
tabla_recomendada = pd.DataFrame(tabla_recomendada)

In [228]:
aps = [] # lista vacía para ir almacenando la AP de cada recomendación


for pred, label in tabla_recomendada[["content_id", "Popular"]].values:
  n = len(pred) # cantidad de elementos recomendados
  arange = np.arange(n, dtype=np.int32) + 1. # indexamos en base 1 
  rel_k = np.in1d(pred[:n], label) # lista de booleanos que indican la relevancia de cada ítem
  tp = np.ones(rel_k.sum(), dtype=np.int32).cumsum() # lista con el contador de verdaderos positivos
  denom = arange[rel_k] # posiciones donde se encuentran los ítems relantes
  ap = (tp / denom).sum() / len(label) # average precision
  aps.append(ap)

In [229]:
MAP = np.mean(aps)
print(f'mean average precision = {round(MAP, 5)}')

mean average precision = 0.00921


In [230]:
MiRec=[]
for i in tabla_recomendada['account_id']:
  recomendaciones=Recomendador(i)
  MiRec.append(recomendaciones)


In [232]:
len(MiRec)

86878

In [233]:
tablafinal=tabla_recomendada.copy()

In [234]:
tablafinal['Mi_recomendacion']=MiRec

In [241]:
aps = [] # lista vacía para ir almacenando la AP de cada recomendación


for pred, label in tablafinal[["content_id", 'Mi_recomendacion']].values:
  n = len(pred) # cantidad de elementos recomendados
  arange = np.arange(n, dtype=np.int32) + 1. # indexamos en base 1 
  rel_k = np.in1d(pred[:n], label) # lista de booleanos que indican la relevancia de cada ítem
  tp = np.ones(rel_k.sum(), dtype=np.int32).cumsum() # lista con el contador de verdaderos positivos
  denom = arange[rel_k] # posiciones donde se encuentran los ítems relantes
  ap = (tp / denom).sum() / len(label) # average precision
  aps.append(ap)

In [242]:
MAP = np.mean(aps)
print(f'mean average precision = {round(MAP, 5)}')

mean average precision = 0.02018
